packages/contracts-bedrock/test/setup/Upgrade.s.sol

```diff
@@ -0,0 +1,193 @@
+        cmds[8] = "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80";
+
+        bytes memory result = Process.run(cmds);
+        return result;
+    }
+
+    /// @notice Executes the op-deployer apply command
+    function opDeployerApply() internal returns (bytes memory) {
+        string[] memory cmds = new string[](8);
+        cmds[0] = "scripts/op-deployer";
+        cmds[1] = "apply";
+        cmds[2] = "--l1-rpc-url";
+        cmds[3] = "http://localhost:8545";
+        cmds[4] = "--private-key";
+        // Private key for first default hardhat account
+        cmds[5] = "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80";
+        cmds[6] = "--workdir";
+        cmds[7] = "test/fixtures";
+
+        bytes memory result = Process.run(cmds);
+        return result;
+    }
+
+    /// @notice Deploy all of the L1 contracts necessary for a full Superchain with a single Op Chain.
+    function run() public {
+        // Note: For now we are using a fork of mainnet, so we don't need to deploy the Superchain Shared contracts.
+        //       This will change after the next release when we have all implementations in a single release.
+
+        // ----- Deploy the OPCM along with implementations contracts it requires -----
+        opDeployerBootstrap();
+        opDeployerApply();
+
+        console.log("Upgrade: reading file state.json");
+        try vm.readFile("test/fixtures/state.json") returns (string memory data_) {
+            _stateJson = data_;
+        } catch {
+            require(false, "Upgrade: cannot find state.json");
+        }
+
+        // Save superchain shared contracts
+        save("SuperchainConfig", _stateJson.readAddress("$.superchainDeployment.superchainConfigImplAddress"));
+        save("SuperchainConfigProxy", _stateJson.readAddress("$.superchainDeployment.superchainConfigProxyAddress"));
+        save("ProtocolVersions", _stateJson.readAddress("$.superchainDeployment.protocolVersionsImplAddress"));
+        save("ProtocolVersionsProxy", _stateJson.readAddress("$.superchainDeployment.protocolVersionsProxyAddress"));
+        save("SuperchainProxyAdmin", _stateJson.readAddress("$.superchainDeployment.proxyAdminAddress"));
+
+        console.log("Upgrade: reading file %s", "test/fixtures/standard-versions.toml");
+        try vm.readFile("test/fixtures/standard-versions.toml") returns (string memory data_) {
+            _stdVersionsToml = data_;
+        } catch {
+            require(false, "Upgrade: cannot find standard-versions.toml");
+        }
+
+        vm.parseToml(_stdVersionsToml, ".releases[\"op-contracts/v1.6.0\"]");
+        // Save OPCM contracts
+        save(
+            "L1CrossDomainMessenger",
+            vm.parseTomlAddress(
+                _stdVersionsToml,
+                string.concat(".releases[\"op-contracts/v1.6.0\"].l1_cross_domain_messenger.implementation_address")
+            )
+        );
```

Comment:
Could we instead just read the OPCM address from the superchain registry, and find all other implementation addresses from that? Generally it's safe to trust the superchain registry, but if we can remove the trust and just read the chain directly, it seems we may as well